feat(checker): add apache http support

gotlougit opened the pull request:
This PR adds the checker for Apache HTTP server, which was requested in issue #1384. This is my first contribution to this project, and I have tried to adhere to the guidelines outlined. Hope the code is satisfactory and would like some feedback on how to improve the checker.

XDRAGON2002 commented on cve_bin_tool/checkers/apache.py:
```diff
@@ -0,0 +1,22 @@
+# Copyright (C) 2021 Intel Corporation
```
This should be 2022 I believe?

BreadGenie reviewed (changes requested):
You need to add tests for the checker. Refer [this guide](https://github.com/intel/cve-bin-tool/tree/main/cve_bin_tool/checkers#adding-tests)

gotlougit commented:
I have added the tests and changed the copyright header as well now.

BreadGenie commented:
@gotlougit you need to run `LONG_TESTS=1 pytest -k <checker-name>` to properly test the checker and to generate the condensed test files. You also need to commit those files too.

Also we follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) for PR titles too.

gotlougit commented:
@BreadGenie I ran the command you told, but I didn't see any condensed test files generated anywhere. The checker seems to pass the test by the way, and I'll keep the PR title in mind from now on. Thanks!

XDRAGON2002 commented:
> @BreadGenie I ran the command you told, but I didn't see any condensed test files generated anywhere. The checker seems to pass the test by the way, and I'll keep the PR title in mind from now on. Thanks!

@gotlougit What he was referring to is the addition of the binary files in to the PR, see #1076 for reference about how the binary files have been committed as well.

gotlougit commented:
I think I've added the compressed download now, but when I try to run the test, one of them is failing, saying that "Condensed downloads are not commited". I have added the tarball in the commits now, but am not able to understand where I went wrong here.

BreadGenie commented on cve_bin_tool/checkers/__init__.py:
```diff
@@ -12,6 +12,7 @@
     "Checker",
     "VendorProductPair",
     "accountsservice",
+    "apache",
```
```suggestion
    "apache_http_server",
```
Apache alone is not good as a checker name since there could be other Apache products.

BreadGenie commented on cve_bin_tool/checkers/apache.py:
```diff
@@ -0,0 +1,22 @@
+
+
+"""
+CVE checker for apache
```
```suggestion
CVE checker for Apache HTTP Server
```
You'd also have to change the filename to apache_http_server.py

BreadGenie commented on cve_bin_tool/checkers/apache.py:
```diff
@@ -0,0 +1,22 @@
+from cve_bin_tool.checkers import Checker
+
+
+class ApacheChecker(Checker):
```
```suggestion
class ApacheHttpServerChecker(Checker):
```

BreadGenie commented on cve_bin_tool/checkers/apache.py:
```diff
@@ -0,0 +1,22 @@
+class ApacheChecker(Checker):
+    CONTAINS_PATTERNS = [
+        r"logs/apache_runtime_status",
+        r"200 OK",
```
```suggestion
```
This pattern can be found in other packages too.

BreadGenie commented:
now if you run `LONG_TESTS=1 pytest -k apache_http_server` the condensed test files will be produced.

BreadGenie commented:
It seems like the commit messages of some commits don't follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/). You might want to squash all the commits into one with proper commit message after the PR is complete.

BreadGenie commented on cve_bin_tool/checkers/apache.py:
```diff
@@ -0,0 +1,21 @@
+    ]
+    FILENAME_PATTTERN = [r"httpd"]
+    VERSION_PATTERNS = [
+        r"Apache/([0-9]+\.[0-9]+(\.[0-9]+)?)",
```
On 2nd look this might not be a unique pattern. This pattern could possibly exist in other Apache products and could lead to false positives.

gotlougit replied:
I found this one:
```httpd-2.4.52-1.fc35.x86_64.debug```

I don't know how Fedora-specific this is though, maybe Ubuntu etc have a similar package name embedded in the binary?

gotlougit replied:
Perhaps we can also see other patterns in the string to make sure it's httpd specifically and not any other Apache product?

BreadGenie replied:
Patterns ending with `.debug` are often considered not optimal. You can look for multi-line strings that are unique to the package too (like in #1200).

gotlougit replied:
I added more patterns in CONTAINS_PATTERNS to be more HTTP server specific, and I wasn't able to find better versioning information in the strings than the one currently in the code.

terriko commented:
Approving the GitHub Actions workflow to run now.

terriko commented on cve_bin_tool/checkers/apache_http_server.py:
```diff
@@ -0,0 +1,23 @@
+    ]
+    FILENAME_PATTTERN = [r"httpd"]
+    VERSION_PATTERNS = [
+        r"Apache/([0-9]+\.[0-9]+(\.[0-9]+)?)",
```
This seems like it might be a problematic pattern, since apache is a pretty large foundation.  Can you check a few of their other products and make sure this is only used in httpd?

gotlougit replied:
Okay, I'll try to do that, however the problem is that this seemed to be the only line that had the version number displayed cleanly in the binary.

terriko commented on test/test_data/apache_http_server.py:
```diff
@@ -0,0 +1,19 @@
+
+mapping_test_data = [
+    {
+        "product": "httpd",
```
This test is failing:
```
>           assert product in list_products
E           AssertionError: assert 'httpd' in {'http_server'}
```

That's basically saying "we found http_server but you told us to find httpd" 

It looks from your checker above that the product name in the NVD database is http_server, so that's probably what should be here in the "product" field?

terriko replied:
This test is failing:
```
>           assert product in list_products
E           AssertionError: assert 'httpd' in {'http_server'}
```

That's basically saying "we found http_server but you told us to find httpd" 

It looks from your checker above that the product name in the NVD database is http_server, so that's probably what should be here in the "product" field?

terriko commented on test/test_data/apache_http_server.py:
```diff
@@ -0,0 +1,19 @@
+    {
+        "url": "http://mirrors.kernel.org/fedora/releases/35/Everything/x86_64/os/Packages/h/",
+        "package_name": "httpd-2.4.51-2.fc35.rpm",
+        "product": "httpd",
```
As above, this test will fail if there's a mismatch between httpd and http_server

terriko replied:
As above, this test will fail if there's a mismatch between httpd and http_server

terriko commented on cve_bin_tool/checkers/apache_http_server.py:
```diff
@@ -0,0 +1,23 @@
+    VERSION_PATTERNS = [
+        r"Apache/([0-9]+\.[0-9]+(\.[0-9]+)?)",
+    ]
+    VENDOR_PRODUCT = [("apache", "http_server")]
```
Highlighting this string so you can easily find it relative to the tests below that are failing.

terriko replied:
Highlighting this string so you can easily find it relative to the tests below that are failing.

gotlougit commented:
I ran the checker test now and it seems to work, so the tests should pass to the best of my knowledge.

terriko commented:
Updating branch and running CI.  I expect gitlint won't like the PR title though, "checker: " isn't a conventional commits prefix.

gotlougit commented on cve_bin_tool/checkers/apache_http_server.py:
```diff
@@ -0,0 +1,23 @@
+    ]
+    FILENAME_PATTTERN = [r"httpd"]
+    VERSION_PATTERNS = [
+        r"Apache/([0-9]+\.[0-9]+(\.[0-9]+)?)",
```
I checked Apache Log4j but wasn't able to find such a pattern for version number. I just did a simple strings search on the jar files.

b31ngd3v commented:
![image](https://user-images.githubusercontent.com/64733912/166660518-c78c93ab-3ad4-4426-aae5-fd2ef3a27f5f.png)
broken url

terriko commented:
I'm still looking for a bit more research on this one to make sure there won't be false positives before I can merge it.   I opened up an issue describing more precisely what I think we still need to do:

* #1701

terriko commented:
Thanks to @Will-Bo's research here:
* #1701

I'm feeling more like this might be safe to merge.  There's still a reasonable chance of a false positive, but it doesn't look like it's a common pattern in the most popular apache projects.  I'm going to update the PR to main for tests and we'll discuss during our regular meeting.

terriko reviewed (approved):
Finally, it's merging time! thanks so much for your work on this one, and your patience in getting it merged.

terriko commented:
I'm going to update to main and run the tests one more time for safety before merging, but I do intend to get this merged when they pass.

gotlougit commented:
Thank you so much! Glad I could help out on this project.

terriko commented:
Looks like we may need a new condensed download before this will merge cleanly.

gotlougit commented:
I think the problem here is that the condensed download is being saved as "http_server-2.4.51-2.fc35.x86_64.rpm.tar.gz" but we need the file "httpd-2.4.51-2.fc35.x86_64.rpm.tar.gz" from the server.

I am a bit confused on how to solve for this. Any ideas?

b31ngd3v commented:
@gotlougit first delete `test/condensed-downloads/http_server-2.4.51-2.fc35.x86_64.rpm.tar.gz` this file, then run `LONG_TESTS=1 pytest -k http_server`, it'll generate a new .tar.gz file. then push the changes.

b31ngd3v commented:
@gotlougit can you please delete this file `test/condensed-downloads/http_server-2.4.51-2.fc35.x86_64.rpm.tar.gz`? i don't think this file is needed

terriko commented:
Okay, let's get this merged!
